<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="More about what it does (maybe more than one line)">
<title>a method for evaluating the pathogenicity of missense variants using AlphaFold2 • MOVA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="a method for evaluating the pathogenicity of missense variants using AlphaFold2">
<meta property="og:description" content="More about what it does (maybe more than one line)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">MOVA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9">

<h1 id="mova">MOVA</h1>
<p>a method for evaluating the pathogenicity of missense variants using
AlphaFold2 <br /> MOVA algorithms are described in detail in:<br />
Yuya Hatano, Tomohiko Ishihara, Osamu Onodera. Accuracy of a machine
learning method based on structural and locational information from
AlphaFold2 for predicting the pathogenicity of TARDBP and FUS gene
variants in ALS. bioRxiv 2022.07.07.499092; doi:
https://doi.org/10.1101/2022.07.07.499092
<h2 id="Installation">Installation</h1> 
<li>MOVA has
been confirmed to work with R 4.2.2.</li>
<li>devtools and LowMACA installation
is required.</li></p>
<pre><code>install.packages(&quot;devtools&quot;, dependencies = TRUE)
if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;LowMACA&quot;)</code></pre>
<ul>
<li>To install MOVA, execute the following two lines</li>
</ul>
<pre><code>library(devtools)
install_github(&quot;yuya-hatano/MOVA&quot;)</code></pre>
<hr />
<h2 id="tutorial">Tutorial</h2>
<ul>
<li>To load the package, enter the following in your R session:</li>
</ul>
<pre><code>library(LowMACA)
library(MOVA)</code></pre>
<ul>
<li>Create a csv file from the HGMD data by manually creating four
columns as shown in the figure below.</li>
<li>Specifically, based on HGMD data, list amino acid changes (column:
HGVSprotein), VCF (column: VCF), DM if DM in HGMD, DMp if DM? (column:
possible), Pathogenic if pathogenic but not a target phenotype, and
Target if target phenotype(column: Type) for each variant as shown in
the figure below and save it. <img
src="https://user-images.githubusercontent.com/108056381/226163655-e107e999-751a-4c11-8eb8-a7b7e8afc4a6.png"
alt="readme_pic1" /></li>
<li>Perform the following in R.</li>
</ul>
<pre><code>Hgmd_divide(&quot;created csv file&quot;)</code></pre>
<ul>
<li>Search for target genes in gnomAD v3.1.2.</li>
<li>Select Ensembl canonical transcript. <img
src="https://user-images.githubusercontent.com/108056381/226165662-1cf3d48a-5c20-44a6-a800-8da242bccdb1.png"
alt="readme_pic2" /></li>
<li>Please select only Missense / inframe indel.</li>
<li>Deselect Indels and select SNVs. <img
src="https://user-images.githubusercontent.com/108056381/226166217-7bf41609-458e-4cf6-9717-cfcf4b1bc3d2.png"
alt="readme_pic3" /></li>
<li>Select 'Export variants to CSV' and download. <img
src="https://user-images.githubusercontent.com/108056381/226166553-c767cbeb-64f7-405c-a172-cc2ad332a5df.png"
alt="readme_pic4" /></li>
<li>Manually remove all but the missense variants from the downloaded
file.</li>
<li>Enter the following in your R session:</li>
</ul>
<pre><code>Edit_gnomAD_file(&quot;downloaded (and modified) gnomAD file&quot;, &quot;output file (csv)&quot;)</code></pre>
<ul>
<li>Please download AlphScore_final.tsv from the following URL.</li>
<li><a href="https://zenodo.org/record/6288139">https://zenodo.org/record/6288139</a></li>
<li>Enter the following in your R session:</li>
</ul>
<pre><code>Edit_variant_data(&quot;Path of AlphScore_final.tsv&quot;, &quot;HGMD file created by Hgmd_divide&quot;, &quot;gnomAD file created by Edit_gnomAD_file&quot;, &quot;Uniprot name of target gene&quot;, &quot;Gene name &quot;, &quot;Output file (csv)&quot;)</code></pre>
<ul>
<li>In addition to the output file, "(gene name)_variant.csv" file and "(gene name)_alph.csv" file are output.</li>
<li>"(gene name)_alph.csv" file is an important file that is ultimately needed to create the final predicted value file.</li>
<li>Please download dbNSFP4.3a from the following URL.</li>
<li><a href="http://database.liulab.science/dbNSFP#version">http://database.liulab.science/dbNSFP#version</a></li>
<li>Enter the following in your R session:</li>
</ul>
<pre><code>Edit_polyphen_data(Path of the dbNSFP4.3a_variant file for the Chromosome in which the gene resides, Path of output file by the Edit_variant_data function, Path of &quot;(gene name)_alph.csv&quot; output by the Edit_variant_data function., uniprot name of target protein, Lower limit position of the targeted gene (hg38), Upper limit position of the targeted gene (hg38), Name of output file 1, Name of output file 2)
Edit_final_variant_file(output file2 by Edit_polyphen_data, output file)</code></pre>
<ul>
<li>Edit fasta file that records the amino acid sequence information of
the target protein. The first line should describe &gt;Protein, and the
amino acid sequence should begin on the second line.</li>
<li>Please download pdb file from the following URL.</li>
<li><a href="https://alphafold.ebi.ac.uk/">https://alphafold.ebi.ac.uk/</a></li>
<li>Enter the following in your R session:</li>
</ul>
<pre><code>MOVA(fasta_file_name, protein_name, pdb_file_name, file output by the Edit_final_variant_file function, output file 1 by the Edit_polyphen_data function, phenotype = &quot;Target&quot;)</code></pre>
<ul>
<li>This will result in analysis by MOVA.</li>
<li>MOVA has many other functions. For a detailed description of each
function, please refer to the following URL.</li>
<li><a href="https://yuya-hatano.github.io/MOVA/docs/reference/index.html">https://yuya-hatano.github.io/MOVA/docs/reference/index.html</a></li>
</ul>


  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>© 2023 Yuya Hatano</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MOVA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yuya Hatano <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9048-5985" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Tomohiko Ishihara <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9915-8665" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Osamu Onodera <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3354-5472" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yuya Hatano.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
